modules/dlm/src/main/java/org/elasticsearch/dlm/DataLifecycleService.java

```diff
@@ -331,6 +431,166 @@ public void onFailure(Exception e) {
         });
     }
 
+    /*
+     * This method executes the given force merge request. Before executing the request, it writes a piece of custom metadata in the
+     * cluster state indicating that force merge has begun. And once the request has completed successfully it writes a piece of custom
+     * metadata in the cluster state indicating that the force merge has completed. The listener is notified when the forcemerge request
+     * has completed successfully, or when the forcemerge fails or the write of the begin token to the cluster state fails. The success or
+     * failure of the write of the completed token to the cluster state is only logged.
+     */
+    private void forceMergeIndex(ForceMergeRequest forceMergeRequest, ActionListener<Void> listener) {
+        assert forceMergeRequest.indices() != null && forceMergeRequest.indices().length == 1 : "DLM force merges one index at a time";
+        final String targetIndex = forceMergeRequest.indices()[0];
+        ActionListener<Void> forceMergeRunningListener = new ActionListener<>() {
+            @Override
+            public void onResponse(Void unused) {
+                logger.trace("DLM issues request to force merge index [{}]", targetIndex);
+                client.admin().indices().forceMerge(forceMergeRequest, new ActionListener<>() {
+                    @Override
+                    public void onResponse(ForceMergeResponse forceMergeResponse) {
+                        logger.info("DLM successfully force merged index [{}]", targetIndex);
+                        setForceMergeMarkerComplete(targetIndex);
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        if (e instanceof SnapshotInProgressException) {
+                            logger.info(
+                                "DLM was unable to force merge index [{}] because it's currently being snapshotted. "
+                                    + "Retrying on the next DLM run",
+                                targetIndex
+                            );
+                            // We need to unset the force merge marker in the cluster state so that DLM will try force merging again
+                            unsetForceMergeMarker(targetIndex);
+                        } else {
+                            logger.error(
+                                () -> Strings.format(
+                                    "DLM encountered an error trying to force merge index [%s], and will not retry",
+                                    targetIndex
+                                ),
+                                e
+                            );
```

Comment:
I was following the pattern we have for deletes (retry on SnapshotInProgressException, fail on everything else). I don't know what other errors we might get and whether they're retryable or not. A number of times makes sense.

Reply:
OK I think I misread how it was being used for deletion. It really only impacts whether we log at info (failures due to snapshots are expected) or error level. Either way, we try again the next time DLM runs. That's what this does now.
I have not yet made it so that force merging is only attempted a limited number of times.